contracts/State.sol

```diff
@@ -72,10 +77,17 @@ contract State is OwnableUpgradeable {
         uint256 state
     );
 
+     /**
+     * @dev SMT address
+     */
+    ISmt public smt;
```

Comment:
You've embedded the Smt into the State, however the clients will need to, at least, get SMT proofs from the SMT contract.
It doesn't look good to make a client work directly with the `Smt` contract address. It is better to incapsulate everything behind the `State` contract

Reply:
Also discussed to keep this contracts separately with `onlyWriter` modifier